src/Compilers/CSharp/Portable/Binder/RefSafetyAnalysis.cs

```diff
@@ -0,0 +1,897 @@
+        {
+            var visitor = new RefSafetyAnalysis(
+                compilation,
+                symbol,
+                inUnsafeRegion: InUnsafeMethod(symbol),
+                useUpdatedEscapeRules: symbol.ContainingModule.UseUpdatedEscapeRules,
+                diagnostics);
+            foreach (var initializer in fieldAndPropertyInitializers)
+            {
+                try
+                {
+                    visitor.VisitFieldOrPropertyInitializer(initializer);
+                }
+                catch (CancelledByStackGuardException e)
+                {
+                    e.AddAnError(diagnostics);
+                }
+            }
+        }
+
+        private static bool InUnsafeMethod(Symbol symbol)
+        {
+            if (symbol is SourceMemberMethodSymbol { IsUnsafe: true })
+            {
+                return true;
+            }
+
+            var type = symbol.ContainingType;
+            while (type is { })
+            {
+                var def = type.OriginalDefinition;
+                if (def is SourceMemberContainerTypeSymbol { IsUnsafe: true })
+                {
+                    return true;
+                }
+                type = def.ContainingType;
+            }
+
+            return false;
+        }
+
+        private readonly CSharpCompilation _compilation;
+        private readonly Symbol _symbol;
+        private readonly bool _useUpdatedEscapeRules;
+        private readonly BindingDiagnosticBag _diagnostics;
+        private bool _inUnsafeRegion;
+        private uint _localScopeDepth;
+        private Dictionary<LocalSymbol, (uint RefEscapeScope, uint ValEscapeScope)>? _localEscapeScopes;
+        private Dictionary<BoundValuePlaceholderBase, uint>? _placeholderScopes;
+        private uint _patternInputValEscape;
+
+        private RefSafetyAnalysis(
+            CSharpCompilation compilation,
+            Symbol symbol,
+            bool inUnsafeRegion,
+            bool useUpdatedEscapeRules,
+            BindingDiagnosticBag diagnostics,
+            Dictionary<LocalSymbol, (uint RefEscapeScope, uint ValEscapeScope)>? localEscapeScopes = null)
+        {
+            _compilation = compilation;
+            _symbol = symbol;
+            _useUpdatedEscapeRules = useUpdatedEscapeRules;
+            _diagnostics = diagnostics;
+            _inUnsafeRegion = inUnsafeRegion;
+            // _localScopeDepth is incremented at each block in the method, including the
+            // outermost. To ensure that locals in the outermost block are considered at
+            // the same depth as parameters, _localScopeDepth is initialized to one less.
+            _localScopeDepth = Binder.CurrentMethodScope - 1;
+            _localEscapeScopes = localEscapeScopes;
+        }
+
+        private ref struct LocalScope
```

Comment:
Could you explain why these and similar structs below are ref structs? It looks like the goal is to save/restore visitor state as we enter and exit certain contexts. Therefore, although they don't contain references, they should still only be put on the stack and not into class fields, etc.